### ISSUE TITLE
Anybody tried Amazon EFS as backend storage for ClickHouse?

### ISSUE THREAD

**Issue by ramazanpolat** [ClickHouse/ClickHouse#31113]
## TL;DR: Anybody tried Amazon EFS as backend storage for ClickHouse?

We want to use Amazon EFS as the backend storage (for some reason detailed below), but we don't have any experience and there is no documentation on the Internet.

### Motivation
We have a setup where the storage needs to scale x10 more than CPU power. It is a data-hungry application, but it doesn't need more power when data gets bigger. At the same time, a mediocre server will be enough for our workload. So just one ClickHouse server (and maybe a replica) will be enough for us, in terms of the CPU power. But the problem is, EBS disks are limited to 16 TB (the only option which can scale to 64 TB is too pricey). So it would be nice to have a type of disk in which they are practically unlimited... So Amazon EFS is practically just like that. They are cheaper. The IOPS and throughput automatically scale by used file size. Especially considering the latest enhancements over NFS with v4.1:

![image](https://user-images.githubusercontent.com/3527695/140578881-87ba535c-dc9b-4d84-9a78-ff180ea9a332.png)

It seems using Amazon EFS will be much cheaper. Also, we won't have any concerns about the disk size limit. When we need more size, it will be provided by Amazon EFS, when we need more CPU power, we can increase the instance size which will be enough for us for years. Besides we also don't want to deal with all those ClickHouse cluster things like zookeeper and all those configs, etc.

**Comment by alexey-milovidov** [comment]
I don't have experience using ClickHouse on EFS and did not hear about any use cases.
There are successful use cases with Ceph, GlusterFS and MooseFS, so I will not be surprised if it works with EFS as well.

If you use EFS, ensure that you have different data directories for different servers. Shared data directory won't work.

As an alternative you can plug in multiple 16 TB EBS disks and build RAID-0 with mdadm on top of them.

**Comment by ramazanpolat** [comment]
I just run the `clickhouse-benchmark` on a server that has both EBS and EFS attached.
I know this is nowhere near being an EBS vs EFS comparison but this is all I can do for now.

The server is t3.xlarge.
EBS storage: 1 TiB gp2, 3000 IOPS, 
EFS storage: 1 TiB general-purpose, provisioned at 1000 MiB/s


| ____________ EBS _____________ | > |____________ EFS ____________ | >|________ EFS/EBS ________ |
| ----------- | ------ |  ---- | ---- | ---- |

| cold   | 1st    | 2nd    | \> | cold   | 1st    | 2nd    | \> | cold | 1st  | 2nd  |
| ------ | ------ | ------ | -- | ------ | ------ | ------ | -- | ---- | ---- | ---- |
| 0.001  | 0.002  | 0.001  | \> | 0.001  | 0.001  | 0.001  | \> | x1    | x2    | x1    |
| 0.053  | 0.04   | 0.038  | \> | 0.074  | 0.036  | 0.043  | \> | x0.72 | x1.11 | x0.88 |
| 0.254  | 0.136  | 0.121  | \> | 0.482  | 0.13   | 0.142  | \> | x0.53 | x1.05 | x0.85 |
| 0.524  | 0.194  | 0.174  | \> | 1.623  | 0.205  | 0.209  | \> | x0.32 | x0.95 | x0.83 |
| 0.71   | 0.375  | 0.376  | \> | 1.823  | 0.425  | 0.361  | \> | x0.39 | x0.88 | x1.04 |
| 1.575  | 1.27   | 1.225  | \> | 3.075  | 1.274  | 1.237  | \> | x0.51 | x1    | x0.99 |
| 0.088  | 0.076  | 0.075  | \> | 0.11   | 0.076  | 0.076  | \> | x0.8  | x1    | x0.99 |
| 0.082  | 0.031  | 0.032  | \> | 0.152  | 0.039  | 0.037  | \> | x0.54 | x0.79 | x0.86 |
| 2.326  | 1.976  | 1.973  | \> | 3.837  | 1.944  | 1.959  | \> | x0.61 | x1.02 | x1.01 |
| 2.751  | 2.244  | 2.183  | \> | 4.251  | 2.517  | 2.168  | \> | x0.65 | x0.89 | x1.01 |
| 0.953  | 0.704  | 0.669  | \> | 2.232  | 0.835  | 0.711  | \> | x0.43 | x0.84 | x0.94 |
| 1.284  | 0.831  | 0.771  | \> | 2.263  | 0.936  | 0.802  | \> | x0.57 | x0.89 | x0.96 |
| 3.409  | 2.972  | 3.072  | \> | 4.947  | 3.078  | 3.026  | \> | x0.69 | x0.97 | x1.02 |
| 4.786  | 4.218  | 4.286  | \> | 7.207  | 4.084  | 4.07   | \> | x0.66 | x1.03 | x1.05 |
| 4.686  | 4.061  | 4.058  | \> | 6.074  | 4.452  | 3.919  | \> | x0.77 | x0.91 | x1.04 |
| 3.539  | 3.287  | 3.191  | \> | 5.349  | 3.191  | 3.12   | \> | x0.66 | x1.03 | x1.02 |
| 12.088 | 11.476 | 11.371 | \> | 14.559 | 11.097 | 11.063 | \> | x0.83 | x1.03 | x1.03 |
| 8.679  | 7.746  | 7.719  | \> | 11.058 | 8.01   | 7.762  | \> | x0.78 | x0.97 | x0.99 |
| 24.023 | 22.915 | 22.783 | \> | 28.227 | 22.692 | 22.459 | \> | x0.85 | x1.01 | x1.01 |
| 0.777  | 0.489  | 0.238  | \> | 1.688  | 0.35   | 0.259  | \> | x0.46 | x1.4  | x0.92 |
| 8.885  | 4.196  | 4.088  | \> | 16.214 | 4.229  | 4.039  | \> | x0.55 | x0.99 | x1.01 |
| 10.116 | 4.387  | 4.215  | \> | 18.284 | 4.339  | 4.233  | \> | x0.55 | x1.01 | x1    |
| 19.825 | 9.816  | 9.468  | \> | 31.926 | 9.819  | 9.475  | \> | x0.62 | x1    | x1    |
| 15.399 | 4.741  | 4.668  | \> | 30.797 | 5.062  | 5.247  | \> | x0.5  | x0.94 | x0.89 |
| 2.502  | 1.261  | 1.121  | \> | 4.744  | 1.308  | 1.122  | \> | x0.53 | x0.96 | x1    |
| 1.358  | 0.972  | 0.982  | \> | 2.785  | 1.022  | 0.957  | \> | x0.49 | x0.95 | x1.03 |
| 1.865  | 1.136  | 1.132  | \> | 4.473  | 1.235  | 1.135  | \> | x0.42 | x0.92 | x1    |
| 8.788  | 3.68   | 3.472  | \> | 15.194 | 3.561  | 3.443  | \> | x0.58 | x1.03 | x1.01 |
| 9.6    | 5.466  | 5.405  | \> | 15.267 | 5.481  | 5.387  | \> | x0.63 | x1    | x1    |
| 4.733  | 4.742  | 4.798  | \> | 5.145  | 4.8    | 4.686  | \> | x0.92 | x0.99 | x1.02 |
| 3.531  | 2.568  | 2.431  | \> | 5.523  | 2.581  | 2.531  | \> | x0.64 | x0.99 | x0.96 |
| 5.217  | 3.306  | 3.327  | \> | 9.329  | 3.263  | 3.452  | \> | x0.56 | x1.01 | x0.96 |
| 21.862 | 21.028 | 20.748 | \> | 25.753 | 20.411 | 20.543 | \> | x0.85 | x1.03 | x1.01 |
| 19.793 | 15.297 | 15.241 | \> | 26.907 | 14.767 | 14.783 | \> | x0.74 | x1.04 | x1.03 |
| 19.678 | 15.219 | 15.316 | \> | 26.541 | 15.051 | 14.838 | \> | x0.74 | x1.01 | x1.03 |
| 5.943  | 5.372  | 5.428  | \> | 6.637  | 5.235  | 5.491  | \> | x0.9  | x1.03 | x0.99 |
| 0.284  | 0.256  | 0.213  | \> | 0.493  | 0.237  | 0.217  | \> | x0.58 | x1.08 | x0.98 |
| 0.102  | 0.085  | 0.1    | \> | 0.171  | 0.095  | 0.094  | \> | x0.6  | x0.89 | x1.06 |
| 0.115  | 0.095  | 0.079  | \> | 0.298  | 0.091  | 0.085  | \> | x0.39 | x1.04 | x0.93 |
| 0.602  | 0.549  | 0.722  | \> | 0.901  | 0.555  | 0.547  | \> | x0.67 | x0.99 | x1.32 |
| 0.063  | 0.023  | 0.023  | \> | 0.171  | 0.042  | 0.035  | \> | x0.37 | x0.55 | x0.66 |
| 0.064  | 0.02   | 0.02   | \> | 0.12   | 0.035  | 0.033  | \> | x0.53 | x0.57 | x0.61 |
| 0.008  | 0.005  | 0.004  | \> | 0.032  | 0.014  | 0.015  | \> | x0.25 | x0.36 | x0.27 |

**Comment by den-crane** [comment]
@ramazanpolat the problem is in ingestion

compare 

```sql
CREATE TABLE default.test_m
(
    A Int64,
    D DateTime,
    B Int64,
    C String
)
ENGINE = MergeTree
PARTITION BY toYYYYMM(D)
ORDER BY (A, B, D)
SETTINGS index_granularity = 8192


clickhouse-benchmark --database default -c 10 <<< "insert into test_m values (1,'2020-01-01 00:00:00', 1, 'x')"
```

**Comment by alexey-milovidov** [comment]
@ramazanpolat That's very interesting observation. Let's add it to https://clickhouse.com/benchmark/hardware/ ?
(my idea is to add infinite number of data points to this page)

So, EFS is about two times slower by throughput.
For the reference, what is the difference in cost?

**Comment by alexey-milovidov** [comment]
> Provisioned Throughput	$6.00 per MB/s

Do you really pay $6000/month for 1GB/sec?

> The baseline rate is 50 MiB/s per TiB of storage

**Comment by ramazanpolat** [comment]
> @ramazanpolat That's very interesting observation. Let's add it to https://clickhouse.com/benchmark/hardware/ ? (my idea is to add infinite number of data points to this page)

Of course, we can add to the benchmark page. I just couldn't find how to upload the results. The page you referenced asks to send results to clickhouse-feedback@yandex-team.com, which I think is obsolete since ClickHouse Inc.

> So, EFS is about two times slower by throughput. For the reference, what is the difference in cost?

It is hard to calculate the price difference since the pricing depends on so many things (IOPS, throughput, volume storage type, etc). But generally, EFS is cheaper than EBS gp3 or even gp2. **Also, Amazon EFS Lifecycle Management automatically moves unused files to the Infrequent Access storage class which makes it much cheaper. Besides, you won't need to wrap your head around Multi-Volume storage config.** 

> > Provisioned Throughput	$6.00 per MB/s
> 
> Do you really pay $6000/month for 1GB/sec?
> 
> > The baseline rate is 50 MiB/s per TiB of storage

Actually, if using baseline rate, throughput is proportional to filesystem size. Considering that you can't specify the storage size when attaching EFS, I either had to fill the volume with TBs of data to catch the EBS performance or had to pick provisioned throughput. So when used in production, 100 TB of data will provide 100x50 = 5000 MiB/s which is quite good compared to EBS which provided 4000 MB/s max for EBS io2 (too expensive). Since the benchmark suit only uses 100GB of data, I just picked the provisioned throughput mode.

**TL;DR**
If you have a high amount of data, EFS is cheaper. If you also have cold data to be served infrequently, that makes EFS the cheapest option.

*Last note*: We can use AWS Data Sync to automatically transfer data from S3 to EFS which solves some integration challenges for us and this is yet another reason to pick EFS.

**Comment by alexey-milovidov** [comment]
Write to feedback@clickhouse.com
And we still receive emails sent to clickhouse-feedback@yandex-team.com
